NDTensors/src/dense/dense.jl

```diff
@@ -3,73 +3,111 @@
-  return Dense(VecT(dim(inds)))
+function Dense{ElR,DataT}(data::AbstractArray) where {ElR,DataT<:AbstractArray}
+  data = convert(DataT, data)
+  return Dense{ElR,DataT}(data)
 end
 
-Dense{ElT}(dim::Integer) where {ElT} = Dense(zeros(ElT, dim))
+# This function is ill-defined. It cannot transform a complex type to real...
+function Dense{ElR}(data::AbstractArray{ElT}) where {ElR,ElT}
+  d = convert(similartype(typeof(data), ElR), data)
+  return Dense{ElR}(d)
+end
 
-Dense{ElT}(::UndefInitializer, dim::Integer) where {ElT} = Dense(Vector{ElT}(undef, dim))
+function Dense{ElT}(data::AbstractArray{ElT}) where {ElT}
+  return Dense{ElT,typeof(data)}(data)
+end
 
-Dense(::Type{ElT}, dim::Integer) where {ElT} = Dense{ElT}(dim)
+function Dense{ElT}(inds::Tuple) where {ElT}
+  return Dense{ElT}(dim(inds))
+end
 
-Dense(x::ElT, dim::Integer) where {ElT<:Number} = Dense(fill(x, dim))
+function Dense{ElT}(dim::Integer) where {ElT}
+  return Dense{ElT,default_datatype(ElT)}(dim)
+end
 
-Dense(dim::Integer) = Dense(Float64, dim)
+Dense{ElT}() where {ElT} = Dense{ElT,default_datatype(ElT)}()
 
-Dense(::Type{ElT}, ::UndefInitializer, dim::Integer) where {ElT} = Dense{ElT}(undef, dim)
+function Dense(data::DataT) where {DataT<:AbstractArray{<:Any,1}}
+  return Dense{eltype(DataT),DataT}(data)
+end
 
-Dense(::UndefInitializer, dim::Integer) = Dense(Float64, undef, dim)
+function Dense(data::DataT) where {DataT<:AbstractArray{<:Any,N}} where {N}
+  #println("Warning: Only vector based datatypes are currenlty supported by Dense. The data structure provided will be vectorized.")
+  vecData = vec(data)
+  return Dense{eltype(vecData),typeof(vecData)}(vecData)
+end
 
-Dense{ElT}() where {ElT} = Dense(ElT[])
-Dense(::Type{ElT}) where {ElT} = Dense{ElT}()
+function Dense(DataT::Type{<:AbstractArray}, dim::Integer)
+  ElT = eltype(DataT)
+  return Dense{ElT,DataT}(dim)
+end
 
-setdata(D::Dense, ndata) = Dense(ndata)
+Dense(ElT::Type{<:Number}, dim::Integer) = Dense{ElT}(dim)
 
-#
-# Random constructors
-#
+function Dense(ElT::Type{<:Number}, ::UndefInitializer, dim::Integer)
+  return Dense{ElT,default_datatype(ElT)}(undef, (dim,))
+end
 
-function randn(::Type{StoreT}, dim::Integer) where {StoreT<:Dense}
-  return Dense(randn(eltype(StoreT), dim))
+function Dense(::UndefInitializer, dim::Integer)
+  datatype = default_datatype()
+  return Dense{eltype(datatype),datatype}(undef, (dim,))
 end
 
+Dense(x::Number, dim::Integer) = Dense(fill!(similar(default_datatype(typeof(x)), dim), x))
+
+Dense(dim::Integer) = Dense(default_eltype(), dim)
+
+Dense(::Type{ElT}) where {ElT} = Dense{ElT}()
+
+## End Dense initializers
+
+setdata(D::Dense, ndata) = Dense(ndata)
+
 copy(D::Dense) = Dense(copy(data(D)))
 
-Base.real(::Type{Dense{ElT,Vector{ElT}}}) where {ElT} = Dense{real(ElT),Vector{real(ElT)}}
+#This is getting closer but is still broken...
+function Base.real(T::Type{Dense})
+  return set_datatype(T, similartype(datatype(T), complex(eltype(T))))
```

Comment:
Looks like this should be `real(eltype(T))`. Also `Type{<:Dense}` in the signature.